SubterfugeCore/Core/Network/NetworkResponse.cs

```diff
@@ -1,19 +1,91 @@
-﻿namespace SubterfugeCore.Core.Network
+﻿using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+
+namespace SubterfugeCore.Core.Network
 {
     /// <summary>
     /// Base Network response. All Network responses should have a 'success' variable to determine if the operation
     /// succeeded and a message describing any issues if it didn't
     /// </summary>
-    public class NetworkResponse
+    public class NetworkResponse<T>
     {
         /// <summary>
-        /// If the network operation was successful
+        /// The HttpResponse object that comes directly from the HttpClient
         /// </summary>
-        public bool Success { get; set; }
+        public HttpResponseMessage HttpResponse { get; set; }
         
         /// <summary>
-        /// Details about the network operation
+        /// If the request resulted in an error, this variable gets populated with the error content.
         /// </summary>
-        public string Message { get; set; }
+        public NetworkError ErrorContent { get; set; }
+        
+        /// <summary>
+        /// The raw string of the HttpResponse. This may be an error string or a valid JSON string.
+        /// </summary>
+        public string ResponseContent { get; set; }
+        
+        /// <summary>
+        /// A templated object that holds the parsed JSON response if the request was successful.
+        /// </summary>
+        public T Response { get; set; }
+        
+        /// <summary>
+        /// Static method to create a new Network Response.
+        /// </summary>
+        /// <param name="responseMessage">The HttpResponseMessage that is recieved from the HttpClient</param>
+        /// <returns>A new instance of a NetworkResponse</returns>
+        public static async Task<NetworkResponse<T>> FromHttpResponse(HttpResponseMessage responseMessage)
+        {
+            NetworkResponse<T> response = new NetworkResponse<T>();
+            response.HttpResponse = responseMessage;
```

Comment:
Shouldn't we impose certain interface constraints on `T`? As it is defined here, the template class could be _anything_ but we are trying to do stuff with the response content obviously, in particular: reading its value and parsing it as a string. From a technical standpoint, that means it should implement a `ToString` style trait or interface.

Reply:
Yep absolutely. I forgot that you could require a generic type to have inheritance restrictions. I will put those in